src/shogun/classifier/LDA.cpp

```diff
@@ -63,28 +50,28 @@ void LDA::init()
 	    SG_OPTIONS(AUTO_LDA, SVD_LDA, FLD_LDA))
 }
 
-LDA::~LDA()
-{
-}
-
 template <typename ST, typename U>
-bool LDA::train_machine_templated(const std::shared_ptr<DenseFeatures<ST>>& data)
+bool LDA::train_machine_templated(
+    const std::shared_ptr<DenseFeatures<ST>>& data,
+    const std::shared_ptr<Labels>& labs)
 {
 	index_t num_feat = data->get_num_features();
 	index_t num_vec = data->get_num_vectors();
 
 	bool lda_more_efficient = (m_method == AUTO_LDA && num_vec <= num_feat);
 
 	if (m_method == SVD_LDA || lda_more_efficient)
-		return solver_svd<ST>(data);
+		return solver_svd<ST>(data, labs);
 	else
-		return solver_classic<ST>(data);
+		return solver_classic<ST>(data, labs);
 }
 
 template <typename ST>
-bool LDA::solver_svd(std::shared_ptr<DenseFeatures<ST>> data)
+bool LDA::solver_svd(
+    std::shared_ptr<DenseFeatures<ST>> data,
+    const std::shared_ptr<Labels>& labs)
 {
-	auto labels = multiclass_labels(m_labels);
+	auto labels = multiclass_labels(labs);
```

Comment:
these are the types of calls that would be removed, as `labs` could be assumed to be multiclass labels (converted by your label encoder in the base class)
I think you could do that in this PR already ... (see comment in Machine::train)

Reply:
@LiuYuHui let me know if you need me to rebase to have access to LabelEncoder

Reply:
yes, it is a great help if you can help me to rebase.